Fix pod stuck in termination state when mount fails or gets skipped after kubelet restart

gnufied opened the pull request:
This PR partially uses changes from - https://github.com/kubernetes/kubernetes/pull/110575 


And fixes https://github.com/kubernetes/kubernetes/issues/96635 and https://github.com/kubernetes/kubernetes/issues/70044 

This PR:

1. Keeps tracks of volumes which were skipped during reconstruction with the assumption that since they were present in DSOW, they will make it to ASOW. 
2. Enhances the check that makes volumes uncertain on unmounted - https://github.com/kubernetes/kubernetes/pull/110670/files#diff-3e8f8a6e5eca57db8e84bcc6c896f0a3b10289fe55d8e999b4bdf1cfdc0de394L785 to ensure that, we are not accidentally marking volumes which were added via reconsturction as unmounted because mount is failing after kubelet restart.

There are two important things in this PR:
1. It adds volumes skipped during reconstruction as uncertain in ASOW.
2. But above change alone is not enough to fix the bug (or unmount the volume), because - when kubelet will retry to mount the volume, mount will fail and it will mark the volume we added as uncertain in previous step as "unmounted" - https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/util/operationexecutor/operation_generator.go#L788 . So basically if mount operation fails, kubelet will undo our hard work of adding volumes which we added as uncertain in step#1.

Now it is clear that - we need to add volumes we read from disk as uncertain, so as kubelet can unmount them properly. But there are two ways to fix the second problem (i.e where kubelet is marking mount point as unmounted, if mount fails for a volume which was marked uncertain before).

1. In this PR - I am keeping track of such reconstructed volumes in ASOW, so as even if mount fails for the volume it *must* not mark the volume was unmounted, if volume was added via reconstruction.
2. There is another easier way - which is to make all volumes uncertain in ASOW if mount fails. That is what - https://github.com/kubernetes/kubernetes/pull/110650 PR does. This code however breaks an existing behaviour where if kubelet starts mounting a volume and may be operation timed-out (marked uncertain) and kubelet retries and mount fails then volume is marked as unmounted and a pod that was using this volume can be deleted without unmounting the volume. If we treat every mount failure as uncertain mount, then volume *must* be unmounted for pod to be deleted.

```release-note
Unmount volumes correctly for reconstructed volumes even if mount operation fails after kubelet restart
```

gnufied committed:
Keep track of each pod that uses a volume during reconstruction

Add tests for volume cleaning up

gnufied committed:
Fix code to process volumes which were skipped during reconstruction

jsafrane commented on pkg/volume/util/operationexecutor/operation_executor.go:
```diff
@@ -215,6 +215,10 @@ type ActualStateOfWorldMounterUpdater interface {
 	// MarkForInUseExpansionError marks the volume to have in-use error during expansion.
 	// volume expansion must not be retried for this volume
 	MarkForInUseExpansionError(volumeName v1.UniqueVolumeName)
+
+	AddVolumeViaReconstruction(opts MarkVolumeOpts) error
+
+	IsVolumeReconstructed(volumeName v1.UniqueVolumeName, podName volumetypes.UniquePodName) bool
```
Missing godoc

jsafrane commented on pkg/volume/util/operationexecutor/operation_generator.go:
```diff
@@ -784,7 +784,8 @@ func (og *operationGenerator) markDeviceErrorState(volumeToMount VolumeToMount,
 
 func (og *operationGenerator) markVolumeErrorState(volumeToMount VolumeToMount, markOpts MarkVolumeOpts, mountError error, actualStateOfWorld ActualStateOfWorldMounterUpdater) {
 	if volumetypes.IsOperationFinishedError(mountError) &&
-		actualStateOfWorld.GetVolumeMountState(volumeToMount.VolumeName, markOpts.PodName) == VolumeMountUncertain {
+		actualStateOfWorld.GetVolumeMountState(volumeToMount.VolumeName, markOpts.PodName) == VolumeMountUncertain &&
+		!actualStateOfWorld.IsVolumeReconstructed(volumeToMount.VolumeName, volumeToMount.PodName) {
```
It would be nice to log some "keeping volume uncertain because it's reconstructed", so we can debug any issues there.

jsafrane commented on pkg/kubelet/volumemanager/reconciler/reconciler.go:
```diff
@@ -249,6 +255,63 @@ func (rc *reconciler) mountAttachedVolumes(volumeToMount cache.VolumeToMount, po
+		}
+
+		if uncertainVolumeCount > 0 {
+			// If the volume has device to mount, we mark its device as mounted.
```
"mark its device as uncertain"

gnufied committed:
Address review comments

gnufied committed:
Add unit test for verifying if processReconstructedVolumes works as expected

gnufied commented on pkg/kubelet/volumemanager/reconciler/reconciler.go:
```diff
@@ -608,69 +701,76 @@ func (rc *reconciler) updateDevicePath(volumesNeedUpdate map[v1.UniqueVolumeName
 	return nil
 }
 
+func (rc *reconciler) markVolumeState(volume *reconstructedVolume, volumeState operationexecutor.VolumeMountState) error {
```
good catch. I also added a unit test for new stuff.

jingxu97 commented:
How about during reconstruction, we put all reconstructed volumes into actual state with uncertain mode even if desired state has this information (we might need to sync between desired state and actual state which already there)?

gnufied commented:
@jingxu97 we could and that is what https://github.com/kubernetes/kubernetes/pull/108180 PR by @jsafrane  does. But adding *all* volumes to ASW in uncertain state during reconstruction brings new challenges, such as if we add a volume to ASW before it gets added to DSW, then we could accidently unmount the volume. So it has all these additional checks to safeguard against similar failure modes - https://github.com/kubernetes/kubernetes/pull/108180/files#diff-2acbd2d6c088f99f9c7e9042b3449418b4824675d2692350c332d0f276489268R40 ( @jsafrane correct me if I am wrong). 

But that alone will not fix the issue, because we still need to avoid marking volumes which we added via recontruction to ASW as "unmounted" (from previous "uncertan" state if mount fails) - https://github.com/kubernetes/kubernetes/pull/110670/commits/f291291be86c1e376624f78601e290acfd3515f5#diff-3e8f8a6e5eca57db8e84bcc6c896f0a3b10289fe55d8e999b4bdf1cfdc0de394R787

So we have this middle ground PR of mine, which avoids those problems and still fixes issues we set out to fix.

gnufied commented:
/retest

gnufied committed:
Remove volume from found during reconstruction if mounted

Add unit tests for removing reconstructed volumes from ASOW

jsafrane commented:
As Hemant pointed out. will need *all* reconstructed volumes in ASW for my SELinux work. But I can have it behind a feature gate + we don't need to worry about backporting it as a bugfix to older releases. (And yes, I need to update it with the latest Hemant's fixes from this PR.) This PR can be somewhere in middle and could be backported to older releases.

gnufied commented:
Yeah eventually once jan's PR gets merged and enabled by default - we could delete the separate `processReconstructedVolumes` function and have reconstruction add all volumes. We will still need part#2 of this PR - which tracks reconstructed volumes in ASOW, so as we are not accidently flipping mount state of uncertain volumes which were added to ASOW via reconstruction.

gnufied commented on pkg/kubelet/volumemanager/cache/actual_state_of_world.go:
```diff
@@ -346,6 +348,40 @@ func (asw *actualStateOfWorld) MarkVolumeAsDetached(
+}
+
 func (asw *actualStateOfWorld) MarkVolumeAsMounted(markVolumeOpts operationexecutor.MarkVolumeOpts) error {
 	return asw.AddPodToVolume(markVolumeOpts)
```
fixed. Also added more tests that verify ASOW accurately.

jsafrane commented:
/kind bug
/lgtm
/hold
But I want @jingxu97's approval / lgtm too.

jingxu97 commented on pkg/kubelet/volumemanager/cache/actual_state_of_world.go:
```diff
@@ -226,7 +227,8 @@ type actualStateOfWorld struct {
 	// containing more information about the attached volume.
-	attachedVolumes map[v1.UniqueVolumeName]attachedVolume
+	attachedVolumes           map[v1.UniqueVolumeName]attachedVolume
+	foundDuringReconstruction map[v1.UniqueVolumeName]map[volumetypes.UniquePodName]types.UID
```
add comments about foundDuringReconstruction?

jingxu97 commented on pkg/kubelet/volumemanager/reconciler/reconciler.go:
```diff
@@ -175,6 +177,10 @@ func (rc *reconciler) reconcile() {
 	// Ensure devices that should be detached/unmounted are detached/unmounted.
 	rc.unmountDetachDevices()
+
+	if len(rc.skippedDuringReconstruction) > 0 {
```
add a comment it should be only executed once?

jingxu97 commented on pkg/kubelet/volumemanager/reconciler/reconciler.go:
```diff
@@ -249,6 +255,74 @@ func (rc *reconciler) mountAttachedVolumes(volumeToMount cache.VolumeToMount, po
+		// check if volume is marked as attached to the node
+		// for now lets only process volumes which are at least known as attached to the node
+		// this should help with most volume types (including secret, configmap etc)
+		if !rc.actualStateOfWorld.VolumeExists(volumeName) {
```
so we might still miss some cases that volumes that are attachable plugin, if here it is not marked as attached yet, and later mount failed?

gnufied replied:
I think if volume was attachable and node still holds the status in `volumeAttached` then - it should not be a problem. Could there be a case where volume is not even attached to the node and still requires cleanup? I think if someone force detached the volume then yeah we won't be cleaning the volume. I think it seems like a very edge case.

jingxu97 replied:
if kubelet restarts, when volume will be marked as exist (attached) in actual state?

gnufied replied:
I have changed the code in `processReconstructedVolumes` to wait for volume to be marked as attached via `VerifyControllerAttached` function. This means that `processReconstructedVolumes` will be called as long as one or more volumes remain in `skippedDuringReconstruction` map.

jingxu97 replied:
could you point me where the wait logic?

gnufied replied:
The wait logic is - the volume is only removed from `skippedDuringReconstruction` if it exists in ASOW.  So for volumes which were added to `skippedDuringReconstruction` during reconstruction and if volume does not exist (not even verified as attached) then volume is not removed from `skippedDuringReconstruction` and `processReconstructedVolumes` is called again for them.

jingxu97 replied:
but when volume will be verified as attached? Only volume is in desired state already? If volume cannot be placed in desired state, it will not added as attached?

jingxu97 replied:
volume can be in desired state temporarily, but it might be removed quickly, and volume is not added in desired state.

jingxu97 commented on pkg/kubelet/volumemanager/reconciler/reconciler.go:
```diff
@@ -366,19 +440,36 @@ type reconstructedVolume struct {
 	blockVolumeMapper   volumepkg.BlockVolumeMapper
 }
 
+type globalVolumeInfo struct {
```
why it is called globalVolumeInfo? maybe add comment?

gnufied replied:
added

jingxu97 commented on pkg/kubelet/volumemanager/reconciler/reconciler.go:
```diff
@@ -608,69 +712,76 @@ func (rc *reconciler) updateDevicePath(volumesNeedUpdate map[v1.UniqueVolumeName
 				continue
 			}
-			klog.V(4).InfoS("Volume is marked device as mounted and added into the actual state", "pod", klog.KObj(volume.pod), "podName", volume.podName, "volumeName", volume.volumeName)
+			klog.V(4).InfoS("Volume is marked device as mounted and added into the actual state", "volumeName", gvl.volumeName)
```
why remove pod info in the log?

gnufied replied:
I removed pod info because here we are making volume as mounted globally and hence pod info does not matter much. But I can add it back.

gnufied replied:
err - also I do not have pod information in `gvl` variable, so that was another reason to remove the pod info. I do not have pod info in gvl because it is `gvl.podVolumes[]` which contains individual pod information. Previously this code was kinda buggy and now it begs the question - if we are to add pod information for global volume mount, then which pod information should we print here? May be just the first one?

jingxu97 replied:
will the pod information ever be logged anywhere?

gnufied replied:
yes pod information gets logged when *volume* is marked as mounted:

```
klog.V(4).InfoS("Volume is marked as mounted and added into the actual state", "pod", klog.KObj(volume.pod), "podName", volume.podName, "volumeName", volume.volumeName)
```

jingxu97 commented on pkg/kubelet/volumemanager/reconciler/reconciler.go:
```diff
@@ -249,6 +255,74 @@ func (rc *reconciler) mountAttachedVolumes(volumeToMount cache.VolumeToMount, po
+					VolumeSpec:          volume.volumeSpec,
+					VolumeMountState:    operationexecutor.VolumeMountUncertain,
+				}
+				err := rc.actualStateOfWorld.AddVolumeViaReconstruction(markVolumeOpts)
```
there might be a race? when you check actual state previously, podvolume is not in actual state, but when you add volume next, it is possible that volume is added into the state.

gnufied replied:
Yes there is a small window where `AddVolumeViaReconstruction` could race with `MarkVolumeAsMounted` or something similar operation. I have thought about it and if `AddVolumeViaReconstruction` is last function to run then mount operation will start again (because we marked something as uncertain which was previously marked as mounted).  I think it should be fine since mount operation is anyways idempotent.

But - it is also possible to fix this race with somewhat heavy handed approach to locking. If we held lock for `PodRemovedFromVolume` and `AddVolumeViaReconstruction` both functions with ever releasing it(say we made a single function that does `CheckAndMarkAsUncertain`), then I believe race is eliminated because either this function will wait for lock or other function will wait for lock. 

I personally do not mind leaving the code as it is, but I am happy to change if you think it is worth fixing for race condition. what do you think?

jingxu97 replied:
Will another race occur?
1. desired state still has the volume since actual state does not has it yet, so it is in the process of mounting
2. volume is added into the actual state through the logic added here.
3. desired state populator checks the actual state has the volume already, and delete volume from desired state
4. the pending mount operation from 1 failed, and mark volume as unmounted.

in this case, pod will not stuck in terminating any more, but mount is not cleaned up?

gnufied replied:
> the pending mount operation from 1 failed, and mark volume as unmounted.

If volume was added to ASOW via reconstruction then a failed mount won't mark volume as unmounted. It will still remain in uncertain state and volume must be unmounted for pod to be deleted.

jingxu97 replied:
oh, right, I forgot you changed the logic there too.

gnufied replied:
I replaced above function with `CheckAndMarkVolumeAsUncertainViaReconstruction`.

jingxu97 commented:
I am thinking again the logic of reconstruction. We still allow desired state populated first, and let reconciler execute without delay. During reconstruction, we can CheckandMarkAsUncertain only if no state in ASW. During reconcile, if mount fail, it should not clean up metadata and markasunmounted if it is uncertain or already mounted. (I am checking old PR https://github.com/kubernetes/kubernetes/pull/82492 to understand why we markvolumeasunmounted. See https://github.com/kubernetes/kubernetes/pull/82492#pullrequestreview-327200934, but that is markdeviceasuncertain) So this way we can avoid interference between reconstruction and reconcile.

gnufied commented:
> During reconcile, if mount fail, it should not clean up metadata and markasunmounted if it is uncertain or already mounted. (I am checking old PR https://github.com/kubernetes/kubernetes/pull/82492 to understand why we markvolumeasunmounted. See https://github.com/kubernetes/kubernetes/pull/82492#pullrequestreview-327200934, but that is markdeviceasuncertain) So this way we can avoid interference between reconstruction and reconcile.

Basically the underlying reason for marking volume as unmounted when mount fails is backward compatibility. Before - https://github.com/kubernetes/kubernetes/pull/82492 patch landed, if mount operation fails then pod+volume basically will not even exist in ASOW. After that change - since pod+volume could exist in ASOW even if mount operation timedout, we made a change to mark the volume as *unmounted* if mount operation fails.

If we don't do that anymore - we will be making a change that is not entirely backward compatible. Whether this behaviour change breaks anyone or not - is kind of hard to predict. This is similar to code that we are removing which checks if path is a mount point. Generally it is safer to do but we are not sure if it will break anyone. 

I see your point of doing `CheckandMarkAsUncertain` during reconstruction and simplifying the ASOW code by not keeping track of reconstructed volumes separately.  My vote is lets revisit this topic via a change that we don't have to backport. We make it 1.26 or something and let it soak during the release cycle. We could even do a release note and develop CSI sanity tests that verify that `NodeUnpublish` should succeed for volumes which were *never* `NodePublished`. 

There is one more thing though - calling `CheckandMarkAsUncertain` during recontruction will not be exact equivalent of code  proposed in this PR.  Because there are other minor differences that needs to be sorted out:

For example - the code in `processReconstructedVolumes` only handles volumes which were marked as attached. This is generally safer to do because, if we also start doing `CheckandMarkAsUncertain` for every volume not found in ASOW during reconstruction, we could be marking device path as wrong, since presence of api-server is not guaranteed during reconstruction. `processReconstructedVolumes` avoids this problem by not processing volumes which were not already marked as attached.

jingxu97 commented:
> > During reconcile, if mount fail, it should not clean up metadata and markasunmounted if it is uncertain or already mounted. (I am checking old PR #82492 to understand why we markvolumeasunmounted. See [#82492 (review)](https://github.com/kubernetes/kubernetes/pull/82492#pullrequestreview-327200934), but that is markdeviceasuncertain) So this way we can avoid interference between reconstruction and reconcile.
> 
> Basically the underlying reason for marking volume as unmounted when mount fails is backward compatibility. Before - #82492 patch landed, if mount operation fails then pod+volume basically will not even exist in ASOW. After that change - since pod+volume could exist in ASOW even if mount operation timedout, we made a change to mark the volume as _unmounted_ if mount operation fails.
> 
> If we don't do that anymore - we will be making a change that is not entirely backward compatible. Whether this behaviour change breaks anyone or not - is kind of hard to predict. This is similar to code that we are removing which checks if path is a mount point. Generally it is safer to do but we are not sure if it will break anyone.
> 
> I see your point of doing `CheckandMarkAsUncertain` during reconstruction and simplifying the ASOW code by not keeping track of reconstructed volumes separately. My vote is lets revisit this topic via a change that we don't have to backport. We make it 1.26 or something and let it soak during the release cycle. We could even do a release note and develop CSI sanity tests that verify that `NodeUnpublish` should succeed for volumes which were _never_ `NodePublished`.
> 
> There is one more thing though - calling `CheckandMarkAsUncertain` during recontruction will not be exact equivalent of code proposed in this PR. Because there are other minor differences that needs to be sorted out:
> 
> For example - the code in `processReconstructedVolumes` only handles volumes which were marked as attached. This is generally safer to do because, if we also start doing `CheckandMarkAsUncertain` for every volume not found in ASOW during reconstruction, we could be marking device path as wrong, since presence of api-server is not guaranteed during reconstruction. `processReconstructedVolumes` avoids this problem by not processing volumes which were not already marked as attached.

I basically agree with you. Just to be honest, I lost track of some code history. Seems to me that before PR #82492, if mount failure does not result in marking volume unmounted, the ASW should still have volume information if it is already there (e.g., remount case).

jingxu97 commented on pkg/volume/util/operationexecutor/operation_generator.go:
```diff
@@ -785,11 +785,19 @@ func (og *operationGenerator) markDeviceErrorState(volumeToMount VolumeToMount,
 		actualStateOfWorld.GetVolumeMountState(volumeToMount.VolumeName, markOpts.PodName) == VolumeMountUncertain {
+		// if volume was previously reconstructed we are not going to change its state as unmounted even
+		// if mount operation fails.
+		if actualStateOfWorld.IsVolumeReconstructed(volumeToMount.VolumeName, volumeToMount.PodName) {
```
will this situation occur? 

1. volume is mounted
2. it triggered remount, but secret was deleted accidentally
3. volume is marked as unmount?

gnufied committed:
Use CheckAndMarkAsUncertainViaReconstruction for uncertain volumes

Also only remove volumes from skippedDuringReconstruction only if
volume was marked as attached.

gnufied commented on pkg/kubelet/volumemanager/cache/actual_state_of_world_test.go:
```diff
@@ -458,6 +460,149 @@ func Test_AddTwoPodsToVolume_Positive(t *testing.T) {
 	verifyVolumeMountedElsewhere(t, podName2, generatedVolumeName2, true /*expectedMountedElsewhere */, asw)
 }
 
+func TestActualStateOfWorld_FoundDuringReconstruction(t *testing.T) {
```
added

gnufied commented:
> Seems to me that before PR https://github.com/kubernetes/kubernetes/pull/82492, if mount failure does not result in marking volume unmounted, the ASW should still have volume information if it is already there (e.g., remount case).

Before https://github.com/kubernetes/kubernetes/pull/82492 any mount failure used to return the call without touching anything in ASOW. So if mount failed (and volume was never mounted successfully previously) then no information about volume  mount gets recorded in ASOW.

For remount case - similarly if mount failed, then since call returns without touching ASOW, it will consider volume as mounted (but we do this today as well, so no change in behaviour).

gnufied commented:
/retest

jingxu97 commented on pkg/kubelet/volumemanager/cache/actual_state_of_world_test.go:
```diff
@@ -458,6 +460,151 @@ func Test_AddTwoPodsToVolume_Positive(t *testing.T) {
+				VolumeSpec:          volumeSpec1,
+				VolumeMountState:    operationexecutor.VolumeMountUncertain,
+			}
+			_, err = asw.CheckAndMarkVolumeAsUncertainViaReconstruction(markVolumeOpts1)
```
maybe return bool value and log it?

gnufied replied:
not sure if it will cause unnecessary noise in test output. Were you thinking of some specific case?

jingxu97 replied:
just want to have some log for the new logic we added. will it affect test output?

jingxu97 replied:
oh, it is in test file, nvm

jingxu97 commented on pkg/kubelet/volumemanager/cache/actual_state_of_world.go:
```diff
@@ -346,6 +350,110 @@ func (asw *actualStateOfWorld) MarkVolumeAsDetached(
+	asw.Lock()
+	defer asw.Unlock()
+
+	volumeObj, volumeExists := asw.attachedVolumes[opts.VolumeName]
```
if it is not in attachdVolumes, what does it mean? In what situation it might happen?

gnufied replied:
This is complementary to logic in reconciler. `CheckAndMarkVolumeAsUncertainViaReconstruction` should only get called for volumes which are considered attached.

And hence if it does get called for volumes which are not attached at all then we are not marking them as mounted in uncertain state.

jingxu97 commented on pkg/kubelet/volumemanager/cache/actual_state_of_world.go:
```diff
@@ -346,6 +350,110 @@ func (asw *actualStateOfWorld) MarkVolumeAsDetached(
+	podObj.volumeMountStateForPod = opts.VolumeMountState
+
+	// if volume is mounted successfully, then it should be removed from foundDuringReconstruction map
+	if opts.VolumeMountState == operationexecutor.VolumeMounted {
```
for now, I see opts.VolumeMountState should always be uncertain, right?

gnufied replied:
yep. hardcoded the state in this function to be `VolumeMountUncertain`.

jingxu97 commented on pkg/kubelet/volumemanager/cache/actual_state_of_world.go:
```diff
@@ -346,6 +350,110 @@ func (asw *actualStateOfWorld) MarkVolumeAsDetached(
+	if mounter != nil {
+		// The mounter stored in the object may have old information,
+		// use the newest one.
+		podObj.mounter = mounter
```
how about we always use the newest information from opts passed no matter podObj exist or not?

gnufied replied:
okay updated to use opts always

jingxu97 commented on pkg/kubelet/volumemanager/reconciler/reconciler_test.go:
```diff
@@ -2181,3 +2185,212 @@ func Test_Run_Positive_VolumeMountControllerAttachEnabledRace(t *testing.T) {
+		verifyFunc           func(rcInstance *reconciler, fakePlugin *volumetesting.FakeVolumePlugin) error
+	}{
+		{
+			name: "when two pods are using same volume and both are deleted",
```
how about two pods are using same volume, but one deleted?

gnufied replied:
added an additional test for this.

jingxu97 commented:
/lgtm

Thank you very much for the great effort on fixing it!

gnufied commented:
/retest

jsafrane commented:
/lgtm

gnufied commented:
/hold cancel

rphillips commented:
Created a cherry pick for 1.24 [here](https://github.com/kubernetes/kubernetes/pull/111636).
